source/validate_cfg.cpp

```diff
@@ -24,32 +24,314 @@
+using std::vector;
+
+using libspirv::BasicBlock;
+
 namespace libspirv {
 
-// TODO(umar): Support for merge instructions
-// TODO(umar): Structured control flow checks
+namespace {
+
+using bb_ptr = BasicBlock*;
+using cbb_ptr = const BasicBlock*;
+using bb_iter = vector<BasicBlock*>::const_iterator;
+
+template <typename T>
+bool Contains(const vector<T>& vec, T val) {
+  return find(begin(vec), end(vec), val) != end(vec);
+}
+
+/// @brief Sorts the blocks in a CFG given the entry node
+///
+/// Returns a vector of basic block pointers in a Control Flow Graph(CFG) which
+/// are sorted in the order they were accessed in a post order traversal.
+///
+/// @param[in] entry the first block of a CFG
+/// @param[in] depth_hint a hint about the depth of the CFG
+///
+/// @return A vector of pointers in the order they were access in a post order
+/// traversal
+vector<const BasicBlock*> PostOrderSort(const BasicBlock& entry, size_t size) {
+  struct block_info {
+    cbb_ptr block;
+    bb_iter current_successor;
+    bb_iter end_successor;
+  };
+
+  vector<cbb_ptr> out;
+  vector<block_info> staged;
+  vector<uint32_t> processed;
+
+  staged.reserve(size);
+  staged.emplace_back(block_info{&entry, begin(entry.get_successors()),
+                                 end(entry.get_successors())});
+  processed.push_back(entry.get_id());
+
+  while (!staged.empty()) {
+    block_info& top = staged.back();
+    if (top.current_successor == top.end_successor) {
+      // No children left to process
+      out.push_back(top.block);
+      staged.pop_back();
+    } else {
+      bb_iter& child_iter = top.current_successor;
+
+      if (!Contains(processed, (*child_iter)->get_id())) {
+        // Process next child
+        staged.emplace_back(block_info{*child_iter,
+                                       begin((*child_iter)->get_successors()),
+                                       end((*child_iter)->get_successors())});
+        processed.push_back((*child_iter)->get_id());
+      }
+      child_iter++;
+    }
+  }
+  return out;
+}
+}
+
+vector<pair<BasicBlock*, BasicBlock*>> CalculateDominators(
+    const BasicBlock& first_block) {
+  struct block_detail {
+    size_t dominator;
+    size_t postorder_index;
+  };
+
+  vector<cbb_ptr> postorder = PostOrderSort(first_block, 10);
+  const size_t undefined_dom = static_cast<uint32_t>(postorder.size());
+
+  unordered_map<cbb_ptr, block_detail> idoms;
+  for (size_t i = 0; i < postorder.size(); i++) {
+    idoms[postorder[i]] = {undefined_dom, i};
+  }
+
+  idoms[postorder.back()].dominator = idoms[postorder.back()].postorder_index;
+
+  bool changed = true;
+  while (changed) {
+    changed = false;
+    for (auto b = postorder.rbegin() + 1; b != postorder.rend(); b++) {
+      // printf("processing: %d\n", (*b)->get_id());
+      size_t& b_dom = idoms[*b].dominator;
+      const vector<BasicBlock*>& predecessors = (*b)->get_predecessors();
+
+      // first processed predecessor
+      BasicBlock* idom =
```

Comment:
Is this guaranteed not to return `end(predecessors)`?  If so, there should be an assertion.